FMFA-418: Add BIGIP GTM DNS Listener module

andreykashcheev opened the pull request:
### Summary 

This change is intended to add GTM DNS Listener module. 


### Changes

 - Add new bigip_gtm_dns_listener module 
 - Add unit test for new module
 - Add functional tests for new module

### Functional Test Results

wojtek0806 commented:
write tests make sure CI/CD passes before i review this

andreykashcheev commented:
Func Tests Results:

PLAY RECAP **************************************************************************************************************************************************************************************************************************************************************
bigip_v12                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v13                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v14                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v15                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

andreykashcheev commented:
@wojtek0806 

> fix the ci/cd failures and write tests before we begin

I am not sure how to fix the following error:
https://travis-ci.com/github/F5Networks/f5-ansible/jobs/330787641

Also, after looking at the F5 SDK, it appears that create/update/delete actions are not permitted: 
https://f5-sdk.readthedocs.io/en/latest/_modules/f5/bigip/tm/gtm/listener.html

I have added func tests results; This MR is ready for review. Please take a look when you have a minute

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+      - enables or disables port translation.
+    type: str
+    default: enabled
+  vlans_enabled:
+    description:
+      - when C(yes), traffic will be accepted by this listener only on the VLANS specified in the vlans option.
+      - only C(yes) is allowed value
+    type: bool
+  vlans_disabled:
+    description:
+      - when C(yes), traffic will not be accepted by this listener only on the VLANS specified in the vlans option.
+      - only C(yes) is allowed value
+    type: bool
+  vlans:
+    description:
+      - specifies a list of VLANs on which traffic is either disabled or enabled.
+      - enabling/disabling decision is based on C(vlans_disabled) and C(vlans_enabled) settings.
+      - removing/reseting VLANs required providing empyt list C([]) as parameter value
+    type: list
+    elements: str
```
can this endpoint support vlans enabled and disabled at the same time, just like VS? if yes then redo this completely

andreykashcheev replied:
In the API doc, it states the following: "Specifies a list of VLANs on which traffic is either disabled or enabled, based on the vlans-disabled and vlans-enabled settings." 
https://clouddocs.f5.com/api/icontrol-rest/APIRef_tm_gtm_listener.html

Testing shows that vlans_enabled and vlans_disabled are mutually exclusive. BIGIP prefers vlans_disabled in a case when both flags provided.

Also, looking at VirtualServer module, it appears that in a case when both provided, VS module takes only disabled vlans: 
https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_virtual_server.py#L2090 


In my implementation, there are three fields, such as: 
   - vlans_enabled  - flag which indicates what to do; mutually exclusive with vlans_disabled
   - vlans_disabled - flag which indicates what to do; mutually exclusive with vlans_enabled. 
   - vlans - list of vlans. Setting to empty list "[]" will make all option 

Let me know if you have  any questions/concerns

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+  description: specifies if address translation is enabled.
+  returned: changed
+  type: str
+  sample: enabled
+translate_port:
+  description: specifies if port translation is enabled.
+  returned: changed
+  type: str
+  sample: enabled
+fallback_persistence:
+  description: fallback persistence profile for the listener to use when the default persistence profile is not available.
+  returned: changed
+  type: str
+  sample: /Common/fallback-profile
+enabled:
+  description: provides DNS listener state.
+  returned: changed
+  type: bool
+  sample: yes
+ip_protocol:
+  description: ip protocol used by DNS listener.
+  returned: changed
+  type: string
+  sample: tcp
+vlan_enabled:
+  description: indicates if traffic will be accepted from specified vlans.
+  returned: changed
+  type: bool
+  sample: no
+vlans:
+  description: list of vlans used by DNS Listener
+  returned: changed
+  type: list
+  sample: ['/Common/vlan1', '/Common/vlan2']
+rules:
+  description: list of rules run by DNS Listener
+  returned: changed
+  type: list
+  sample: ['/Common/rule1', '/Common/rule2']
+'''
```
capital letters and full stops

andreykashcheev replied:
Updated all to upper-case and add missing fullstop. 

This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+            return ''
+        return self._values['description']
+
+    @property
+    def address(self):
+        if self._values['address'] is None:
+            return None
+        return self._values['address']
+
+    @property
+    def port(self):
+        if self._values['port'] is None:
+            return None
+        return self._values['port']
+
+    @property
+    def ip_protocol(self):
+        if self._values['ip_protocol'] is None:
+            return None
+        return self._values['ip_protocol']
+
+    @property
+    def mask(self):
+        if self._values['mask'] is None:
+            return None
+        return self._values['mask']
+
+    @property
+    def translate_address(self):
+        if self._values['translate_address'] is None:
+            return None
+        return self._values['translate_address']
+
+    @property
+    def translate_port(self):
+        if self._values['translate_port'] is None:
+            return None
+        return self._values['translate_port']
```
why are those here?

andreykashcheev replied:
I agree - since we are not transforming those parameters values, defining them here are unnecessary. I have removed all properties which do not require transformation from ModuleParameter class. 

This comment is addressed now. Let me know if you have any questions/concerns

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+        return self._values['translate_port']
+
+    @property
+    def advertise(self):
+        if self._values['advertise'] is None:
+            return None
+        return flatten_boolean(self._values['advertise'])
+
+    @property
+    def vlans_enabled(self):
+        if self._values['vlans_enabled'] is None:
+            return None
+        elif self._values['vlans_enabled'] is True:
+            return True
+        return False
+
+    @property
+    def vlans_disabled(self):
+        if self._values['vlans_disabled'] is None:
+            return None
+        elif self._values['vlans_disabled'] is True:
+            return True
+        return False
```
incorrect

andreykashcheev replied:
Thanks for pointing out to the existent logic in virtual server module! 
I have implemented the same way.
This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+        return False
+
+    @property
+    def vlans(self):
+        if self._values['vlans'] is None:
+            return None
+        if is_empty_list(self._values['vlans']):
+            return []
+        result = list()
+        for item in self._values['vlans']:
+            result.append(item)
+        return result
```
why do you append list to a list? what is the purpose

andreykashcheev replied:
I agree - this is unnecessary for loop and it was removed to return just vlans value. 

Thanks for pointing out to the existent logic in virtual server module! 

I have implemented the same way as it is done for virtual server. 

This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+            results.append(result)
+        return results
+
+    @property
+    def last_hop_pool(self):
+        if self._values['last_hop_pool'] is None:
+            return None
+        return self._values['last_hop_pool']
+
+    @property
+    def source_port(self):
+        if self._values['source_port'] is None:
+            return None
+        return self._values['source_port']
+
+    @property
+    def fallback_persistence(self):
+        if self._values['fallback_persistence'] is None:
+            return None
+        return self._values['fallback_persistence']
+
+    @property
+    def auto_lasthop(self):
+        if self._values['auto_lasthop'] is None:
+            return None
+        return self._values['auto_lasthop']
```
why are those here?

andreykashcheev replied:
I agree - since no transformation is done on those properties, defining them here is unnecessary. 

This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+        return result
+
+    @property
+    def rules(self):
+        if self._values['rules'] is None:
+            return None
+        if is_empty_list(self._values['rules']):
+            return ''
+        results = list()
+        for irule in self._values['rules']:
+            result = fq_name(self.partition, irule)
+            results.append(result)
+        return results
```
why are you returning empty string to the API if this is a list? does the endpoint accept empty list?

andreykashcheev replied:
Thanks for pointing out to the existent logic in virtual server module! 

I have implemented the same way as it is done for virtual server. 

This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+
+
+class ReportableChanges(Changes):
+    @property
+    def address(self):
+        return self._values['address']
+
+    @property
+    def mask(self):
+        return self._values['mask']
+
+    @property
+    def port(self):
+        return self._values['port']
+
+    @property
+    def vlans(self):
+        return self._values['vlans']
+
+    @property
+    def rules(self):
+        return self._values['rules']
```
why are those here ?

andreykashcheev replied:
I agree - those are unnecessary since no transformation happens with their values

Removed them all. This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    def vlans_disabled(self):
+        return flatten_boolean(self._values['vlans_disabled'])
+
+    @property
+    def advertise(self):
+        if self._values['advertise'] is None:
+            return None
+        return flatten_boolean(self._values['advertise'])
+
+    @property
+    def ip_protocol(self):
+        if self._values['ip_protocol'] is None:
+            return None
+        return self._values['ip_protocol']
```
incorrect

andreykashcheev replied:
This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+            )
+
+    @property
+    def vlans(self):
+        if self.want.vlans is None:
+            return None
+        elif self.want.vlans == [] and self.have.vlans is None:
+            return None
+        elif self.want.vlans == self.have.vlans:
+            return None
+        else:
+            return self.want.vlans
+
+    @property
+    def rules(self):
+        if self.want.rules is None:
+            return None
+        elif self.want.rules == [] and self.have.rules is None:
+            return None
+        elif self.want.rules == self.have.rules:
+            return None
+        else:
+            return self.want.rules
```
change this look at other modules for examples, also rules does not return empty list but empty string so you have a bug here

andreykashcheev replied:
Thanks for pointing out the existent logic! 

I have implemented the same way as it is done for virtual server. 

This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+
+    def update_on_device(self):
+        params = self.changes.api_params()
+        if 'address' not in params:
+            params['address'] = self.have.address
```
if the address is required then this is not needed

andreykashcheev replied:
Copy that - I removed this if condition and mark address parameter as required under ArgumentSpec 

This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+            address=dict(),
+            port=dict(type='int'),
+            advertise=dict(type='bool'),
+            vlans_disabled=dict(type='bool', choices=[True]),
+            vlans_enabled=dict(type='bool', choices=[True]),
+            vlans=dict(type='list', elements='str'),
```
these can be done differently look at virtual server

andreykashcheev replied:
Thanks for pointing out the existent logic! 

I have implemented the same way as it is done for virtual server. 

This comment is addressed now.

wojtek0806 commented on test/integration/bigip_gtm_dns_listener.yaml:
```diff
@@ -0,0 +1,52 @@
+
+  vars:
+    __metadata__:
+      upstream: no
```
once its finished set yes

andreykashcheev replied:
This is addressed now

wojtek0806 commented on test/integration/targets/bigip_gtm_dns_listener/tasks/main.yaml:
```diff
@@ -0,0 +1,195 @@
+    that:
+        - result is changed
+        - result is success
+        - result.enabled
```
this needs to change

andreykashcheev replied:
This comment is addressed now

wojtek0806 commented:
also update your devel and rebase before submitting changes

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+      - disabled
+extends_documentation_fragment: f5networks.f5_modules.f5
+author:
+  - Andrey Kashcheev (@andreykashcheev)
```
I have reviewed and updated docs strings using instructions provided in the following document:
https://docs.f5net.com/display/~wypior/Chapter+3+-+Code+Conventions#Chapter3-CodeConventions-WritingDocumentation

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    description:
+      - enables or disables port translation.
+    type: str
+    default: enabled
```
This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    default: enabled
+  translate_port:
+    description:
+      - enables or disables port translation.
```
This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    default: preserve
+  translate_address:
+    description:
+      - enables or disables address translation for the listener.
```
This is addressed now.

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    description:
+      - enables or disables address translation for the listener.
+    type: str
+    default: enabled
```
Updated to boolean and removed default.

This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    description:
+      - specifies whether the system preserves the source port of the connection.
+    type: str
+    default: preserve
```
Removed default value. This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    type: int
+  source_port:
+    description:
+      - specifies whether the system preserves the source port of the connection.
```
This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    type: str
+  port:
+    description:
+      - specifies the port on which the listener listens for connections.
```
This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    required: True
+  pool:
+    description:
+      - specifies a default pool to which you want the listener to automatically direct traffic.
```
This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+      - specifies the netmask for a network listener only.
+      - netmask clarifies whether the host bit is an actual zero or a wildcard representation.
+    default: ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
+    required: True
```
No, this was copied over incorrectly while I worked on creating docs. 
Removed incorrectly added required: True 

This comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    required: True
+  address:
+    description:
+      - specifies the IP address on which the system listens.
```
Updated all descriptions to upper case letters

this comment is addressed now

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    description:
+      - specifies the netmask for a network listener only.
+      - netmask clarifies whether the host bit is an actual zero or a wildcard representation.
+    default: ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
```
Removed default value, which was copied over for API doc: 
https://clouddocs.f5.com/api/icontrol-rest/APIRef_tm_gtm_listener.html

This comment is addressed now.

andreykashcheev commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,836 @@
+    description:
+      - Provides brief description for DNS Listener.
+    type: str
+    required: True
```
Removed unnecessary required: True.  

This review is addressed now

andreykashcheev commented:
Func Tests results after addressing reviews:

PLAY RECAP **************************************************************************************************************************************************************************************************************************************************************
bigip_v12                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v13                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v14                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v15                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright: (c) 2019, F5 Networks Inc.
```
2020

andreykashcheev replied:
I have updated the module and unit test with the correct year as well as I have updated the stub files used for generating modules and unit tests. 

This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+      - irule4
+    module: ltm
+    name: '{{ item }}'
+    provider:
+      password: secret
+      server: lb.mydomain.com
+      user: admin
+- name: 'Create iRules'
+  bigip_gtm_dns_listener:
+    address: '192.0.1.0'
+    advertise: false
+    auto_lasthop: default
+    description: 'this is a test DNS listener'
+    enabled_vlans:
+      - /Common/external
+    ip_protocol: tcp
+    irules:
+      - /Common/irule1
+    mask: '255.255.255.0'
+    name: test-dns-listener
+    port: 30025
+    provider:
+      password: secret
+      server: lb.mydomain.com
+      user: admin
+    source_port: preserve
+    state: present
+    translate_address: true
+    translate_port: true
+  delegate_to: localhost
+- name: 'Disable a DNS Listener'
+  bigip_gtm_dns_listener:
+    address: '192.0.1.0'
+    state: disabled
+    name: test-dns-listener
+    provider:
+      password: secret
+      server: lb.mydomain.com
+      user: admin
+  delegate_to: localhost
```
space these out each task separated by a space, also not needed to include tasks from other modules

andreykashcheev replied:
This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+    description:
+      - Specifies list of iRules to run on listener.
+      - iRules help automate the intercepting, processing, and routing of application traffic.
+      - If you want to remove existing iRules, provide an empty list value; C([]) or empty string value C('').
```
should only be empty list, this will cause issues if you specify a string '' on argument which is a type list, with element of string, it used to work but the behavior changed, ansible expects [''] or [] rather than '', so better to just tell them to specify empty list

andreykashcheev replied:
Copy that - I have updated the comment as suggested.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+            return ''
+        return self._values['description']
+
+    @property
+    def translate_address(self):
+        if self._values['translate_address'] is None:
+            return None
+        if self._values['translate_address']:
+            return 'enabled'
+        else:
+            return 'disabled'
+
+    @property
+    def translate_port(self):
+        if self._values['translate_port'] is None:
+            return None
+        if self._values['translate_port']:
+            return 'enabled'
+        else:
+            return 'disabled'
```
why aren't these booleans? flatten_boolean?

andreykashcheev replied:
This is done the same way as in other modules:
https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_virtual_server.py#L2175

Both properties (translate_address and translate_port) are defined as booleans however, BIGIP REST API requires enabled/disabled values; therefore, corresponding parameters return enabled/disabled strings

Link to API Doc: 
https://clouddocs.f5.com/api/icontrol-rest/APIRef_tm_gtm_listener.html

wojtek0806 replied:
incorrect, this is the proper way to do this, change it:

https://github.com/F5Networks/f5-ansible/blob/devel/library/modules/bigip_cgnat_lsn_pool.py#L571

wojtek0806 replied:
booleans in Ansible can be anything, flatten boolean is there for a reason

andreykashcheev replied:
Copy that - I have the change as suggested. 
This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+    def vlans_enabled(self):
+        if self._values['enabled_vlans'] is None:
+            return None
+        elif self._values['vlans_enabled'] is False:
+            # This is a special case for 'all' enabled VLANs
+            return False
```
where is this set?

andreykashcheev replied:
You are correct! This else if is a dead-code which never will be executed since vlans_enabled is not provided as parameter to module and instead, vlans_enabled/vlans_disabled gets populated based on enabled_vlans/disabled_vlans values. 

I have removed unnecessary  else if statement. This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+    def vlans_disabled(self):
+        if self._values['disabled_vlans'] is None:
+            return None
+        elif self._values['vlans_disabled'] is True:
+            # This is a special case for 'all' enabled VLANs
+            return True
```
where is this set?

andreykashcheev replied:
You are correct! This else if is a dead-code which never will be executed since vlans_enabled is not provided as parameter to module and instead, vlans_enabled/vlans_disabled gets populated based on enabled_vlans/disabled_vlans values. 

I have removed unnecessary  else if statement. This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+        if self._values['irules'] is None:
+            return None
+        if is_empty_list(self._values['irules']):
+            return ''
```
return empty list []

andreykashcheev replied:
This comment is addressed now.

wojtek0806 replied:
return empty list

andreykashcheev replied:
I have the change. This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+        return self._values['vlans']
+
+    @property
+    def irules(self):
+        if self._values['irules'] is None:
+            return None
+        if self._values['irules'] == '':
+            return []
+        return self._values['irules']
```
remove this once you address the comment for irules in ModuleParameters

andreykashcheev replied:
This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+        if len(self._values['vlans']) > 0 and self._values['vlans_disabled'] is True:
+            return self._values['vlans']
+
+    @property
+    def irules(self):
+        if self._values['irules'] is None:
+            return None
+        if len(self._values['irules']) == 0:
+            return []
+        if len(self._values['irules']) == 1 and self._values['irules'][0] == '':
+            return ''
+        return self._values['irules']
```
change this once you address the comment in ModuleParameters

andreykashcheev replied:
This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+                version=warning['version']
+            )
+
+    def _override_vlans_if_all_specified(self):
+        '''Overrides any specified VLANs if 'all' VLANs are specified
+
+        The special setting 'all VLANs' in a BIG-IP requires that no other VLANs
+        be specified. If you specify any number of VLANs, AND include the 'all'
+        VLAN, this method will erase all of the other VLANs and only return the
+        'all' VLAN.
+        '''
+        all_vlans = ['/common/all', 'all']
+        if self.want.enabled_vlans is not None:
+            if any(x for x in self.want.enabled_vlans if x.lower() in all_vlans):
+                self.want.update(
+                    dict(
+                        enabled_vlans=[],
+                        vlans_disabled=True,
+                        vlans_enabled=False
+                    )
+                )
```
this would be better suited to be done in ModuleParameters rather then here?

andreykashcheev replied:
I agree - this is better suited to be done in ModuleParameters. I have made the change and this comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,943 @@
+            self.want.update({'disabled': True})
+        elif self.want.state in ['present', 'enabled']:
+            self.want.update({'enabled': True})
+        self._override_vlans_if_all_specified()
```
remove after addressing above comment

andreykashcheev replied:
This comment is addressed now.

wojtek0806 commented on test/integration/targets/bigip_gtm_dns_listener/tasks/main.yaml:
```diff
@@ -0,0 +1,195 @@
+- name: Enable GTM DNS Listener  - Idempotent check
+  bigip_gtm_dns_listener:
+      name: "{{ listener_name }}"
+      address: 10.0.0.0
+      state: enabled
+  register: result
+
+- name: Assert Enable GTM DNS Listener  - Idempotent check
+  assert:
+    that:
+        - result is not changed
+        - result is success
+
+- name: Remove GTM DNS Listener
+  bigip_gtm_dns_listener:
+      name: "{{ listener_name }}"
+      address: 10.0.0.0
+      state: absent
+  register: result
+
+- name: Assert Remove GTM DNS Listener
+  assert:
+    that:
+        - result is changed
+        - result is success
+
+- name: Remove GTM DNS Listener - Idempotent check
+  bigip_gtm_dns_listener:
+      name: "{{ listener_name }}"
+      address: 10.0.0.0
+      state: absent
+  register: result
+
+- name: Assert Remove GTM DNS Listener - Idempotent check
+  assert:
+      that:
+          - result is not changed
+          - result is success
+
+- import_tasks: teardown.yaml
```
reduce those spaces in indentations

andreykashcheev replied:
This comment is addressed now.

wojtek0806 commented on test/integration/targets/bigip_gtm_dns_listener/tasks/teardown.yaml:
```diff
@@ -0,0 +1,38 @@
+---
+
+- name: Delete Pool
+  bigip_pool:
+      lb_method: round-robin
+      name: web-pool
+      state: absent
+  register: result
+
+- name: Delete vlans
+  bigip_vlan:
+      name: "{{ item }}"
+      state: absent
+  loop:
+      - external
+      - internal
+  register: result
+
+- name: Delete irule
+  bigip_irule:
+      content: "when HTTP_REQUEST {HTTP::respond 200 content 'Good Request'}"
+      module: ltm
+      name: "{{ item }}"
+      state: absent
+  loop:
+      - irule1
+      - irule2
+      - irule3
+      - irule4
+  register: result
+
+- name: De-provision GTM
+  bigip_provision:
+      name: gtm
+      state: absent
+  tags:
+      - module-provisioning
+      - deprovision-module
```
too many spaces in indentations

andreykashcheev replied:
This comment is addressed now.

wojtek0806 commented on test/units/modules/network/f5/test_bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,299 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright: (c) 2019, F5 Networks Inc.
```
2020

andreykashcheev replied:
This comment is addressed. I also have updated the stub files.

wojtek0806 commented on test/units/modules/network/f5/test_bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,299 @@
+        mm.read_current_from_device = Mock(return_value=current)
+        results = mm.exec_module()
+
+        assert results['changed'] is True
```
maybe test the 'all' vlan scenario here

andreykashcheev replied:
I have added unit tests to validate enabled_vlans:all and disabled_vlans:all cases
This comment is addressed now

andreykashcheev commented:
@wojtek0806 

I have addressed all comments as well as re-run func tests. Func tests results are provided below: 

PLAY RECAP **************************************************************************************************************************************************************************************************************************************************************
bigip_v12                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v13                  : ok=29   changed=13   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v14                  : ok=29   changed=13   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v15                  : ok=29   changed=13   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,890 @@
+  - BIG-IP DNS listeners allows to use TCP and UDP protocols.
+notes:
+  - Requires BIG-IP version 12.0.0 or greater
+version_added: 2.8
```
wrong version

andreykashcheev replied:
I have updated it to 
```version_added: "f5_modules 1.4"```
 This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,890 @@
+description:
+  - Defines one or more listener objects to control what protocols are available for the BIG-IP DNS system to process DNS requests.
+  - BIG-IP DNS listeners allows to use TCP and UDP protocols.
+notes:
+  - Requires BIG-IP version 12.0.0 or greater
```
remove

andreykashcheev replied:
It is removed. This comment is addressed now

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,890 @@
+            return 'disabled'
+
+    @property
+    def advertise(self):
+        if self._values['advertise'] is None:
+            return None
+        return flatten_boolean(self._values['advertise'])
```
does API require yes/no ? if yes then you can leave it, if not change accordingly

andreykashcheev replied:
Yes - the API sets advertise field to yes/no; by default, it is set to no

```
    "address": "192.0.2.0",
    "advertise": "no",
    "autoLasthop": "default",
```

andreykashcheev replied:
This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,890 @@
+    def irules(self):
+        if self._values['irules'] is None:
+            return None
+        if len(self._values['irules']) == 0:
+            return []
```
if the list is empty just do truth check, empty lists, strings etc, always are False:

`if not self._values['irules']`

andreykashcheev replied:
Right - I have made the change as suggested
This comment is addressed now.

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,890 @@
+    def irules(self):
+        if self.want.irules is None:
+            return None
+        if self.want.irules == '' and len(self.have.irules) > 0:
+            return []
+        if self.want.irules in [[], ''] and len(self.have.irules) == 0:
+            return None
```
change this

andreykashcheev replied:
I have updated as suggested to 
```if not self.want.irules:```

wojtek0806 commented on library/modules/bigip_gtm_dns_listener.py:
```diff
@@ -0,0 +1,890 @@
+        elif any(x for x in self._values['vlans'] if x.lower() in ['/common/all', 'all']):
+            return []
+        return self._values['vlans']
+
```
if you need to return empty string to device for `irules` you do this here, i would check however if the empty list is also accepted, most likely it is accepted so you can ignore it and return empty list in ModuleParameters as per my comment above

andreykashcheev replied:
Correct - the API accepts empty list value; no need to set it to empty string

andreykashcheev commented:
Functional Tests results:

```
PLAY RECAP **************************************************************************************************************************************************************************************************************************************************************
bigip_v12                  : ok=29   changed=9    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v13                  : ok=29   changed=13   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v14                  : ok=29   changed=13   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
bigip_v15                  : ok=29   changed=13   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```
